src/core/thread/mle_router.cpp

```diff
@@ -1461,7 +1462,14 @@ void MleRouter::UpdateRoutes(const RouteTlv &aRoute, uint8_t aRouterId)
 
                 if (neighbor->GetLinkQualityOut() != linkQuality)
                 {
+                    uint8_t oldRouteCost = mRouterTable.GetLinkCost(*neighbor);
                     neighbor->SetLinkQualityOut(linkQuality);
+
+                    // reset MLE advertisement timer if neighbor link cost changed to or from infinite
+                    if ((oldRouteCost >= kMaxRouteCost) != (mRouterTable.GetLinkCost(*neighbor) >= kMaxRouteCost))
```

Comment:
I think this should consider whether a route to the neighbor becomes available or unavailable. Note that it is possible to have a route even if a link has not been established. Rather than simply comparing link cost, we should also see if the route utilizes this link. Thoughts?

Reply:
Agree completely 👍

Reply:
Agree. I will update the code accordingly.

Reply:
I have updated the code:
- when updating neighbor link quality out, reset adv interval only when there is no multihop route to the neighbor (i.e. `mRouterTable.GetRouter(neighbor->GetNextHop()) == NULL`)
- modify `if (cost + mRouterTable.GetLinkCost(*neighbor) <= kMaxRouteCost)` to `if (cost + mRouterTable.GetLinkCost(*neighbor) < kMaxRouteCost)` . 

Per my standing, the neighbor is used as route nexthop only when `route cost < kMaxRouteCost`, and many code in `UpdateRoutes` rely on this restriction. Please verify if there is a bug here.